### DISCUSSION TITLE
Can I make the same OS run on different PCs with exact same speed?

### DISCUSSION THREAD

**Discussion by akayiz** [bochs-emu/Bochs#424] [Q&A]
Hello!

I am studying an operating systems course. The course's style is that we work all semester on an operating systems project in which we add features using C inside an operating system made inside our university and run and test our operating system using Bochs.

There is a big problem that I want to solve which is that the Lecturer tells us to work on functionality X, e.g. let's say X is kernel memory allocation, and then he sends us a `.c` program that we run inside our command-line operating system using Bochs and this program tests the functionality that we added in the OS and shows if the test succeeded or not, the problem is that usually, he wants to ensure we wrote an efficient solution so he sets time constraints for the test, for example, he says the test should take a maximum of 30 seconds, so what we do (as in all students that enrolled in the course) is we run the test and set a timer on google as soon as we run it and see how long it took, BUT the problem that I am trying to solve is that the SAME OS code that I wrote with the SAME test when ran on DIFFERENT computers with different capabilities take different times, so the SAME OS code and test on computer A it can take 20 seconds and on computer B 80 seconds if computer B is a lot slower, so this causes a lot of confusion for students since they don't know if the problem is with their implementation's efficiency or if it's their computers that are slower than the lecturer's computer, since at the end when we submit our final project the lecturer does the tests and grading on his computer.

I was trying to find a solution for this, I played a bit with the Bochs CPU and clock configuration inside `.bochsrc` but I wasn't able to maintain a constant IPS.

I know this may not be 100% related to Bochs. But I would greatly appreciate it if anyone would have an idea on how I can solve this problem, whether through Bochs or writing C code that interacts with Bochs to measure how long a test takes fairly across different computers with different capabilities or any other solution that could help me fix this problem and make our Operating Systems course better!

This is an [old version](https://github.com/mahossam/FOS-Ain-Shams-University-Educational-OS) of the operating system we work on if anyone would like to take a look, it should give a good idea of it since it hasn't changed much.

Thank you for taking the time to read this and thank you in advance for any answers!

**Comment by Vort** [comment]
I have two ideas:
1. Set IPS to more or less high value (for example, `50000000`), `clock: sync=none` and measure time inside of Bochs.
2. Set IPS to small value (for example, `5000000`), `clock: sync=slowdown` and measure time outside of Bochs.

**Reply by akayiz** [reply]
Thank you for your reply, so just to clarify:
 
In both points you mean I try this and then run the same test with the same OS on 2 different PCs and check if it takes the same time.
 
1. How can I measure time inside Bochs?

2. You mean I run the test and then measure time from outside Bochs like with a stopwatch right?

I will make sure to test this as soon as I get the time and let you know how it went.

**Reply by Vort** [reply]
1. I was thinking about modifying test to show time. But another alternative is to look at log file for lines like `00185620000i[SIM   ] quit_sim called with exit code 1`. With `20000000` IPS it means simulation took 185620000 / 20000000 = 9.2 "virtual" seconds.
2. Yes. For example, in Windows, "Date and Time" dialog can be opened with ticking clock. Of course, this task can be automated as well.

**Reply by luibgn** [reply]
Simply set ips to a value less than the max ips value that you can reach on every system.
So if host A can run at X isp and host B can run bochs at Y ips with Y>X because B is a more powerful machine, then choose X. X can be handled by both.
Then set clock sync = slowdown.
